Cython/Compiler/ParseTreeTransforms.py

```diff
@@ -3757,7 +3767,9 @@ def visit_TryFinallyStatNode(self, node):
         """
         Take care of try/finally statements in nogil code sections.
         """
-        if not self.nogil or isinstance(node, Nodes.GILStatNode):
+        if (not self.nogil or
+                isinstance(node, Nodes.GILStatNode) or
+                isinstance(node, Nodes.CriticalSectionStatNode)):
```

Comment:
IMHO, these would better be handled with separate visitor methods (`visit_GILStatNode` etc.) that override the super class handling.